### PR TITLE
[CI] Enforce linting

### PR DESCRIPTION
**Pull request by jkasten2**
# Description
## One Line Summary
Enable enforcement of eslint and prettier rules, as well as made the code changes required them to pass.

## Details
Recommend reviewing commit-by-commit.
The following two commands were run on the source.
* `yarn prettier --write src test __test__`
* `yarn eslint src --ext .js,.jsx,.ts,.tsx --fix`

Since not all rules had auto fixes, some changes were done manually to meet the rules. Also a few in-line suppresses were added to since it was to risky to change the code at this time, but preferred over omitting the rule as the rule should still be applied to new code.

# Validation
Tested on Windows 11 22H2 with Chrome 115.0.5790.171 ensuring it can subscribe, receive a push, register an email, and open a new tab on notification click.
### Checklist
   - [X] All the automated tests pass or I explained why that is not possible
   - [X] I have personally tested this on my machine or explained why that is not possible
   - [X] I have included test coverage for these changes or explained why they are not needed

**Programming Checklist**
Interfaces:
   - [X] Don't use default export
   - [X] New interfaces are in model files

Functions:
   - [X] Don't use default export
   - [X] All function signatures have return types
   - [X] Helpers should not access any data but rather be given the data to operate on.

Typescript:
   - [X] No Typescript warnings
   - [X] Avoid silencing null/undefined warnings with the exclamation point

Other:
   - [X] Iteration: refrain from using `elem of array` syntax. Prefer `forEach` or use `map`
   - [X] Avoid using global OneSignal accessor for `context` if possible. Instead, we can pass it to function/constructor so that we don't call `OneSignal.context`

### Checklist
   - [X] I have included screenshots/recordings of the intended results or explained why they are not needed

---

## Related Tickets

---

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Website-SDK/1103)

### CODE REVIEW

**Review by emawby — APPROVED** [review]
Looks good the tests are passing now in the PR!

File: src/shared/libraries/Emitter.ts [hunk 1]
```diff
@@ -41,10 +41,12 @@ export default class Emitter {
    * execute only once.
    */
   public once(event: string, listener: EventHandler): Emitter {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
     const that = this;
 
-    function fn(this: Function) {
+    function fn(this: unknown) {
```
Comment (rgomezp) [inline src/shared/libraries/Emitter.ts:45]: There are 18 new `unknown` types in this commit. I think we can fix at least some of these such as here

**Review by rgomezp — CHANGES_REQUESTED** [review]

File: src/shared/libraries/Emitter.ts (hunk 1 above)
Comment (jkasten2) [reply src/shared/libraries/Emitter.ts]: This one is fixable, it should be `EventHandler`, fixed it in commit 543f21a59cf5b472a9f9535cc2f3cbeef23beea4.

Most of the other ones are an improvement though still, example `Function` was turninged into `(param: unknown) => void` which is more accurate.
* `Function` is a banned type with the [default eslint rules](https://typescript-eslint.io/rules/ban-types/)

A few of the `unknowns` could probably be replaced with generics, however that will require adding new types which seems out of scope for this PR.

File: src/shared/models/UpdatePlayerOptions.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ export interface UpdatePlayerOptions {
     sdk                         ?: string;
     device_model                ?: string;
     app_id                      ?: string;
-    tags                        ?: Object | null;
+    tags                        ?: Record<string, unknown> | null;
```
Comment (rgomezp) [inline src/shared/models/UpdatePlayerOptions.ts]: what about this record, we know this should be a string, correct?

File: src/shared/api/OneSignalApi.ts [hunk 3]
```diff
@@ -21,15 +21,15 @@ export default class OneSignalApi {
     return OneSignalApiShared.sendNotification(appId, playerIds, titles, contents, url, icon, data, buttons);
   }
 
-  static jsonpLib(url: string, fn: Function) {
+  static jsonpLib(url: string, fn: (err: unknown, data: ServerAppConfig) => void) {
```
Comment (rgomezp) [inline src/shared/api/OneSignalApi.ts]: could this `err` be of type `Error`?

File: src/shared/api/OneSignalApi.ts [hunk 4]
```diff
@@ -21,15 +21,15 @@ export default class OneSignalApi {
...
     JSONP(url, null, fn);
   }
 
   static async downloadServerAppConfig(appId: string): Promise<ServerAppConfig> {
     if (SdkEnvironment.getWindowEnv() !== WindowEnvironmentKind.ServiceWorker) {
       return await new Promise<ServerAppConfig>((resolve, reject) => {
         // Due to CloudFlare's algorithms, the .js extension is required for proper caching. Don't remove it!
-        OneSignalApi.jsonpLib(`${SdkEnvironment.getOneSignalApiUrl().toString()}/sync/${appId}/web`,(err, data) => {
+        OneSignalApi.jsonpLib(`${SdkEnvironment.getOneSignalApiUrl().toString()}/sync/${appId}/web`,(err: unknown, data: ServerAppConfig) => {
```
Comment (rgomezp) [inline src/shared/api/OneSignalApi.ts]: see above

**Review by rgomezp — CHANGES_REQUESTED** [review]

File: src/shared/models/UpdatePlayerOptions.ts (hunk 2 above)
Comment (jkasten2) [reply src/shared/models/UpdatePlayerOptions.ts]: ah yes, good catch! Fixed now

File: src/shared/api/OneSignalApi.ts (hunk 3 above)
Comment (jkasten2) [reply src/shared/api/OneSignalApi.ts]: ah yes, good catch! Fixed now

File: src/shared/api/OneSignalApi.ts (hunk 4 above)
Comment (jkasten2) [reply src/shared/api/OneSignalApi.ts]: ah yes, good catch! Fixed now

**Review by rgomezp — APPROVED** [review]

**Pull request merged**